### ISSUE TITLE
Throwing an exception in an async test does not mean failure

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/e2e#109]
I am not sure whether I misunderstood something about async functions or what's going on, but tests just do not fail when I throw an exception in the navigator object. I need to run some experiments with regular HTML to figure out whether this is a problem with mocha, or it is something async - sync related. Still I cannot throw an unexpected connection error by scripted redirections, so this is something serious.

**Comment by inf3rno** [comment]
#89 is blocked by this for now

**Comment by inf3rno** [comment]
Well I think the problem is even more serious than I expected. The current code does not have a navigation promise, because it looks like this:

```js
        const emptyPage = await navigator.load(`${pages}/empty.html`);
        navigator.window.location.assign(`${pages}/empty.html`);
        await task.defer(200);
```

We need a navigator.started call to create a navigation promise. So the current code just triggers the loading events, throws unexpected navigation expection because there is no promise present, but that does not make the test fail. I need to test this manually in a browser in a reduced testing scenario probably with timeouts. I want to know whether this is a mocha bug or a browser bug or something else.

**Comment by inf3rno** [comment]
this throws exception:
```js
    it("should throw exception", async function  (){
        setTimeout(function (){
            throw new Error("bad");
        },100);
        await new Promise(function (resolve, reject){
            setTimeout(resolve, 200);
        });
    });
```
while this doesn't:

```js
        task.defer(100).then(function (){
            throw new Error("bad");
        });
        await task.defer(200);
```

the only difference is using the .then() so I suspect that is causing the problem

**Comment by inf3rno** [comment]
Yepp, this logs to the console, but the test does not fail because of it:

```js
    it("should throw exception", async function  (){
        new Promise(function (resolve, reject){
            setTimeout(resolve,100);
        }).then(function (){
            console.log("bad");
            throw new Error("bad");
        });
        await new Promise(function (resolve, reject){
            setTimeout(resolve, 200);
        });
    });
```

**Comment by inf3rno** [comment]
We need to check in plain HTML without mocha with more browsers.

**Comment by inf3rno** [comment]
Well this is interesting. We don't really need the second part.

```js
    new Promise(function (resolve, reject){
        setTimeout(resolve,100);
    }).then(function (){
        console.log("a");
        throw new Error("bad");
        console.log("b");
    });
```

This throws error in Chromium based browsers, but not in Firefox. I already sent a bug report. https://bugzilla.mozilla.org/show_bug.cgi?id=1502809

**Comment by inf3rno** [comment]
Okay, this is an unhandled rejection, which is supported by Chrome, but not by the other browsers. I guess Mocha does not work with it either, that's why it cannot see the error thrown.

**Comment by inf3rno** [comment]
I checked again, this is a bug in Firefox and the other browsers. The ".then" returns a new Promise and throwing an Error in that should result in an unhandled rejection, which should be logged, but it isn't. It can be caught with ".catch" and thrown again with "Promise.reject" if needed.

**Comment by inf3rno** [comment]
Since Chrome supports this properly, I am not sure why mocha doesn't.

**Comment by inf3rno** [comment]
I asked it on SO: https://superuser.com/questions/1370930/is-there-a-way-to-convert-an-unhandled-rejection-into-an-uncaught-error but I doubt there will be an answer. I checked, it throws an error if I defer it.

**Comment by inf3rno** [comment]
We can use the following:

```js
            Promise.resolve().then(() => {
                    // ...
                    throw new UnexpectedNavigation();
                    // ...
            }).catch(function (error){
                setTimeout(function (){
                    throw error;
                }, 1);
            });
```

The only problem that it is deferred, so it won't happen in the right time. This might be a problem, I am not sure. I'd rather need a general solution, which works for the task lib too, otherwise the lib will be a huge mess.

**Comment by inf3rno** [comment]
I think what I can do here is the following: `window.addEventListener("unhandledrejection", function (err){throw err;})`.

**Comment by inf3rno** [comment]
This is a Mocha bug after all and it should work normally, so fixing it this way might not be a problem.

**Comment by inf3rno** [comment]
I just need a mocha boot file for the fix. I won't add it to every single test as a module, and the navigator should not contain it either.

**Comment by inf3rno** [comment]
Hmm I might call something on the mocha object if there is one. I need to check the code of mocha.

**Comment by inf3rno** [comment]
Well this is not as easy as I thought. Using the event listener works, but it is still not sync, so the only advantage over setTimeout 1, that I can easily remove it, after they properly implement reporting by unhandled rejections. The only problem that I cannot catch that unhandled rejection for testing purposes. So for example when I test redirection without navigator.start, I expect an unhandled rejection to be present, but there is no way to stop it from triggering an error. I need to check chai as promised maybe that helps. If not, then I need to override the window.addEventListener to fix this.

**Comment by inf3rno** [comment]
Chai as promised does not seem to support this. Another option to use a different microtask defer, so I won't need to care about Promises and everything is fine. Except I still need to catch the error somehow.

**Comment by inf3rno** [comment]
It would be a lot better to solve this with navigtor.error() or something like that.

**Comment by inf3rno** [comment]
With error interface on navigator:
```js
            Promise.resolve().then(() => {
                    // ...
                    throw new UnexpectedNavigation();
                    // ...
            }).catch(function (error){
                    navigator.throw(error);
            });
```

```js
        const emptyPage = await navigator.load(`${pages}/empty.html`);
        navigator.window.location.assign(`${pages}/empty.html`);
        const error = await Promise.race([navigator.catch(), task.defer(200)]);
        expect(error instanceof NavigationError).to.equal(true);
```

If the navigator.catch is not called, then it will throw the error again. We still need to current unhandled rejection fix for mocha, but this would be a lot better than writing an assertion for unhandled rejections. At least I think so. Another advantage, that we could add an event interface instead an report these kind of errors too.

**Comment by inf3rno** [comment]
Another possible workaround:

```js
           await Promise.resolve().then(() => {
                    // ...
                    throw new UnexpectedNavigation();
                    // ...
            });
```

**Comment by inf3rno** [comment]
Ok, the upper does not work, it still results in an unhandled rejection, so we still need the fix.

**Comment by inf3rno** [comment]
I thought about this meanwhile, and I decided to create a promise instead of using the events for handling connection problems. So by every page load a new promise will be created for the next one and I will store the resolve/reject somewhere, so I can call it directly. This solves the unexpected navigation problem too, because I'll have a promise to reject. I'll still need the mocha fix, but that is a mocha dependent hax, so it might not be required by other testing frameworks.

**Comment by inf3rno** [comment]
Hmm I think I'll need to stop the navigation by rejection.

**Comment by inf3rno** [comment]
Meanwhile I wrote a small lib to be able to do something like

```js
const promise = Promise.exposed().then(console.log);
promise.resolve(123);
```

Not that it would make a difference here. But it might make a difference in a test, I am not sure.
